search.py

```diff
@@ -477,7 +477,8 @@ def genetic_algorithm(population, fitness_fn, ngen=1000, pmut=0.1):
     "[Fig. 4.8]"
     for i in range(ngen):
         new_population = []
-        for i in len(population):
+        len_population = len(population)
+        for i in len_population:
```

Comment:
Shouldn't it be for `i in range(len(population))`
Also the length is not calculated in every iteration dynamically. The length is calculated once and used throughout the loop.